Add first draft of nf-core/config specifications (#4087)
nf-core/website: sites/docs/src/content/docs/specifications/configs/maintenance-and-deprecation.md

markdownPlugin: addNumbersToHeadings
weight: 1
---

The keywords "MUST", "MUST NOT", "SHOULD", etc. are to be interpreted as described in [RFC 2119](https://tools.ietf.org/html/rfc2119).

## Maintenance

### Primary maintainer(s) responsibility

The names listed in the `config_profile_contact` SHOULD be responsible for maintaining the config as the primary maintainer.
The primary maintainer MAY be more than one person.

Primary maintainers SHOULD test any changes to configs on their infrastructure.
Primary maintainers SHOULD merge their own pull requests with changes

If a primary maintainers no longer have access to the infrastructure, the maintainer SHOULD make a best effort to transfer the maintainer to another contact person who still has access.

### nf-core community

The nf-core community core- and maintainers teams MAY make minor modifications to configs on the nf-core/configs repository to ensure compatibility with Nextflow and the nf-core template.
The nf-core community core- and maintainers teams MUST NOT make major changes without permission and testing from the primary contact (for example, changing conditionals for selecting queues)

> 

Reviewer: Give us the rights to transfer maintainership

```suggestion
Should the primary maintainers become unresponsive, the nf-core community core- and maintainers teams MAY transfer maintainership to another, active, member of the institution willing to take the responsibility on.

```

Author: Ah nice one!

## Deprecation

### Deprecation through change of primary maintainer

> A primary maintainer MAY only deprecate a config if they are no longer at the institution hosting the infrastructure, if they found no person to transfer maintainership to.

Reviewer: There may be other reasons for someone to not be able a config, e.g. change of role, long leave, etc.
```suggestion
A primary maintainer MAY only deprecate a config if they are no longer able to maintain it and found no person to transfer maintainership to.
```

## Maintenance

### Primary maintainer(s) responsibility

The names listed in the `config_profile_contact` SHOULD be responsible for maintaining the config as the primary maintainer.
The primary maintainer MAY be more than one person.

Primary maintainers SHOULD test any changes to configs on their infrastructure.
Primary maintainers SHOULD merge their own pull requests with changes

If a primary maintainers no longer have access to the infrastructure, the maintainer SHOULD make a best effort to transfer the maintainer to another contact person who still has access.

### nf-core community

The nf-core community core- and maintainers teams MAY make minor modifications to configs on the nf-core/configs repository to ensure compatibility with Nextflow and the nf-core template.
The nf-core community core- and maintainers teams MUST NOT make major changes without permission and testing from the primary contact (for example, changing conditionals for selecting queues)
Should the primary maintainers become unresponsive, the nf-core community core- and maintainers teams MAY transfer maintainership to another, active, member of the institution willing to take the responsibility on.

> ## Deprecation

Reviewer 2: Do we want to allow configs to be left in an "unmaintained" but not deprecated state? What does deprecation actually entail - removal from the repo?

Author: > Do we want to allow configs to be left in an "unmaintained" but not deprecated state? 

I think so, could still be users using it and no changes to the infrastructure even if the person maintaining is no longer there.

> What does deprecation actually entail - removal from the repo?

good point, will add

Reviewer 2: > I think so, could still be users using it and no changes to the infrastructure even if the person maintaining is no longer there.

Can we add a warning to the config (not just in the description), something like:

```
log.warn("WARNING: The {confname} config is unmaintained and may not work as intended on the target infrastructure. If you use this infrastructure, please consider becoming the maintainer or speak to your system administration team about it.")
```

Reviewer 3: I like the idea, but is it possible to access log in a config?

Reviewer 2: ```
params.config_profile_description = {
    def deprecation_message = "WARNING: The {confname} config is unmaintained and may not work as intended on the target infrastructure. If you use this infrastructure, please consider becoming the maintainer or speak to your system administration team about it."
    log.warn(deprecation_message)
    return deprecation_message
}.call()
```

Would this pass nf-schema's type checking?

Reviewer 2: `log.warn` is not available:

ERROR ~ Unknown method invocation `warn` on ConfigObject type -- Did you mean?
  wait

Reviewer 2: I guess we would need to set `params.config_profile_deprecated = true` and then have a section in the template that prints a warning if it's true. That might be overengineering.

Author: I think so, in most cases I imagine people won't be looking anyway if we are indeed deprecarting 😬  (and can always resurrect if necessayr)

Author: Thinking abotu it, it happens so rarely maybe we can do the warning on a case by case basis, and update accordingly :)
